xdsl/dialects/comb.py

```diff
@@ -153,21 +227,115 @@ class XorOp(VariadicCombOperation):
     name = "comb.xor"
 
 
+@dataclass
+class ComparisonOperation:
+    """
+    A generic comparison operation, operation definitions inherit this class.
+
+    The first argument to these comparison operations is the type of comparison
+    being performed, the following comparisons are supported:
+
+    -   equal (mnemonic: `"eq"`; integer value: `0`)
+    -   not equal (mnemonic: `"ne"`; integer value: `1`)
+    -   signed less than (mnemonic: `"slt"`; integer value: `2`)
+    -   signed less than or equal (mnemonic: `"sle"`; integer value: `3`)
+    -   signed greater than (mnemonic: `"sgt"`; integer value: `4`)
+    -   signed greater than or equal (mnemonic: `"sge"`; integer value: `5`)
+    -   unsigned less than (mnemonic: `"ult"`; integer value: `6`)
+    -   unsigned less than or equal (mnemonic: `"ule"`; integer value: `7`)
+    -   unsigned greater than (mnemonic: `"ugt"`; integer value: `8`)
+    -   unsigned greater than or equal (mnemonic: `"uge"`; integer value: `9`)
+    """
+
+    @staticmethod
+    def _get_comparison_predicate(
+        mnemonic: str, comparison_operations: dict[str, int]
+    ) -> int:
+        if mnemonic in comparison_operations:
+            return comparison_operations[mnemonic]
+        else:
+            raise VerifyException(f"Unknown comparison mnemonic: {mnemonic}")
+
+    @staticmethod
+    def _validate_operand_types(operand1: SSAValue, operand2: SSAValue):
+        if operand1.type != operand2.type:
+            raise TypeError(
```

Comment:
```suggestion
            raise ValidationError(
```
TypeError is for an error in our Python runtime world, as opposed to the compilation process. All errors to do with invariants in the IR itself should be ValidationErrors, as we have special logic around those errors specifically.

Reply:
I see, I think that was "lifted off" the `arith` implementation; it's the same error type there.
But I agree `ValidationError` is more fitting.